{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary\n",
    "\n",
    "We would like to take an evolutionary MSA for DHFR and create an energy function following [2013Cocco](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003176). Then we will use this energy function to score single and double mutants from the Homo sapiens version of DHFR. \n",
    "\n",
    "The probability model takes the form [Eqn (7) 2013Cocco](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003176#pcbi.1003176.e059)\n",
    "$$ P(a_1, \\ldots, a_L) = \\frac{1}{Z} \\exp{ \\bigg\\{ \\frac{1}{2} \\sum_{i,j} e_{ij}(a_i, a_j) + \\sum_{i} h_i(a_i) \\bigg\\} }.$$\n",
    "This represents the probability of seeing a given sequence $(a_1, \\ldots, a_L)$ in the MSA. Using the mean field approximation we get values for the coupling matrix $e_{ij}(a_i,a_j)$ and the local fields $h_i(a_i)$. It is hard to calculate the normalization constant $Z$ as $L$ is large. However, since $Z$ is a constant for every sequence in this MSA we can ignore it and get a relative score for each sequence by estimating $-\\log (PZ)$. We define an energy function $\\mathcal{H}(a_1, \\ldots, a_L)$ as $$P(a_1, \\ldots, a_L) = \\exp{ \\big\\{ -\\mathcal{H}(a_1, \\ldots, a_L) \\big\\} }$$\n",
    " \n",
    "**Sections**\n",
    "1. [Import pre-computed data and generate covariance matrix](#gencovmat)\n",
    "2. [Generate probability model parameters](#probparams)\n",
    "3. [Calculate energy for human DHFR](#energycalc)\n",
    "4. [Mutants Energy Calculations](#mutants_energy)\n",
    "5. [Simulate Rounds of Neutral Evolution](#simulate_evo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import pre-computed data and generate covariance matrix <a id=\"gencovmat\" />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import scipy\n",
    "import scipy.linalg\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "AMINO_ACIDS = np.array([aa for aa in \"RKDEQNHSTCYWAILMFVPG-\"], \"S1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'VRPLNCIVAVSQNMGIGKNGDLPWPPLRNEFKYFQRMTTTSSVEGKQNLVIMGRKTWFSIPEKNRPLKDRINIVLSRELKEPPRGAHFLAKSLDDALRLIEQPELASKVDMVWIVGGSSVYQEAMNQPGHLRLFVTRIMQEFESDTFFPEIDLGKYKLLPEYPGVLSEVQEEKGIKYKFEVYEKKD'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the first line of the alignment (which has the WT version of DHFR)\n",
    "datadir = \"../data\"\n",
    "msa_file = os.path.join(datadir, \"DHFR.aln\")\n",
    "with open(msa_file) as fh:\n",
    "    wt_DHFR = np.array([[x for x in fh.readline().strip()] ], np.dtype(\"S1\"))\n",
    "\n",
    "wt_DHFR.tostring().decode('ascii')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading weights from :  ../data/DHFR.weights.npy\n",
      "Loading single site marginals from  ../data/DHFR.single.npy\n",
      "Loading double site marginals from  ../data/DHFR.double.npy\n"
     ]
    }
   ],
   "source": [
    "weights_file = os.path.join(datadir, \"DHFR.weights.npy\")\n",
    "\n",
    "print(\"Loading weights from : \", weights_file)\n",
    "weights = np.load(weights_file)\n",
    "\n",
    "single_site_marginal_file = os.path.join(datadir, \"DHFR.single.npy\")\n",
    "double_site_marginal_file = os.path.join(datadir, \"DHFR.double.npy\")\n",
    "\n",
    "print(\"Loading single site marginals from \", single_site_marginal_file)\n",
    "f_i_a = np.load(single_site_marginal_file)\n",
    "\n",
    "print(\"Loading double site marginals from \", double_site_marginal_file)    \n",
    "f_i_j_a_b = np.load(double_site_marginal_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the length of the sequence and the length of the alphabet from the imported matrices\n",
    "L, q = f_i_a.shape\n",
    "M_eff = sum(weights) # number of effected sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add pseudo count\n",
    "pseudo_count_ratio = 0.5\n",
    "\n",
    "f_i_a = (pseudo_count_ratio / q ) + (1 - pseudo_count_ratio) * f_i_a / M_eff\n",
    "f_i_j_a_b = (pseudo_count_ratio  / (q*q) ) + (1 - pseudo_count_ratio) * f_i_j_a_b / M_eff\n",
    "\n",
    "# The formula for f_i_j_a_b is a little different when i==j\n",
    "# essentially we have f_i_a on the diagonal and zero's everywhere else. \n",
    "for i in range(L):\n",
    "    f_i_j_a_b[i, :, i, :] = np.diag(f_i_a[i, :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C_i_j_a_b.shape = (186, 20, 186, 20)\n"
     ]
    }
   ],
   "source": [
    "# Covariance Matrix\n",
    "# We take an outer product of f_i_a with itself using numpy's broadcasting rules. \n",
    "# This gives us a matrix where the (i,a, j, b) index is f[i,a] * f[j,b]\n",
    "C_i_j_a_b = f_i_j_a_b  - f_i_a[:, :, np.newaxis, np.newaxis] * f_i_a[np.newaxis, np.newaxis, :, :] \n",
    "\n",
    "# we project the covariance matrix down the first q-1 elements\n",
    "# Since the frequencies add up to 1 we can discard amino-acid value (a = q) for each site\n",
    "# without losing any information\n",
    "C_i_j_a_b = C_i_j_a_b[:, :(q-1), :, :(q-1)]\n",
    "print(\"C_i_j_a_b.shape = {}\".format(C_i_j_a_b.shape)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate probability model parameters <a id=\"probparams\" />\n",
    "\n",
    "From [Eqn (8) 2013Cocco](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003176#pcbi.1003176.e071) we have \n",
    "$$e_{ij}(a,b) = -(C^{-1})_{ij}(a,b) \\qquad \\forall i,j, \\enspace \\forall a,b=1, \\ldots, q-1$$ and $$e_{ij}(a,q) = e_{ij}(q,a) = 0 \\qquad \\forall a = 1, \\ldots, q,$$ and $$ h_i(q) = 0.$$\n",
    "\n",
    "It is likely that the negative sign was omitted by accident in Eqn (8)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov = C_i_j_a_b.reshape((L*(q-1), L*(q-1)))\n",
    "invCov = np.linalg.inv(cov)\n",
    "e_i_j_a_b =  - invCov.reshape((L, q-1, L, q-1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the coupling matrix $e_{ij}(a_i, a_j)$, we need to calculate the local fields $h_i(a_i)$. To do this we use [Eqn (S6) from the Supporting Information of 2013Cocco](https://doi.org/10.1371/journal.pcbi.1003176.s001). \n",
    "\n",
    "$$h_i(a) = \\log \\bigg( \\frac{f_i(a)}{f_i(q)} \\bigg) - \\frac{1}{L} \\sum_{\\mu j b} \\xi_i^\\mu(a) \\xi_j^\\mu (b) f_j(b)$$\n",
    "\n",
    "Here we are not distinguishing between positive and negative patterns to lighten the notations. From [Eqn 18 2013 Cocco](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003176#pcbi.1003176.e144) we have $$ e_{ij}(a,b) = \\frac{1}{L} \\sum_\\mu \\xi_i^\\mu (a) \\xi_j^\\mu (b).$$ This gives us\n",
    "$$ \\sum_{jb} e_{ij}(a,b) f_j(b) = \\frac{1}{L} \\sum_{\\mu j b} \\xi_i^\\mu (a) \\xi_j^\\mu (b) f_j(b).$$ Substituting this back into Eqn (S6) we get an expression for the local fields in terms of the coupling matrix and the mariginal frequencies. \n",
    "$$ h_i(a) = \\log \\bigg( \\frac{f_i(a)}{f_i(q)} \\bigg) - \\sum_{jb} e_{ij}(a,b) f_j(b).$$ Here, $b$ runs from $1$ to $q-1$.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(186, 20)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_i_a = np.log(f_i_a[:, :(q-1)] / f_i_a[:, q-1, np.newaxis]) - \\\n",
    "    (e_i_j_a_b * f_i_a[np.newaxis, np.newaxis, :, :(q-1)]).sum(axis=(2,3))\n",
    "\n",
    "h_i_a.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate energy score for WildType DHFR  <a id=\"energycalc\" />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import skbio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Taken from Mus musculus (house mouse) dihydrofolate reductase\n",
    "# https://www.ebi.ac.uk/ena/data/view/AAA37638\n",
    "# the start and stop codons were deleted\n",
    "wt_dna = 'GTTCGACCATTGAACTGCATCGTCGCCGTGTCCCAAAATATGGGGATTGGCAAGAACGGAGACCTACCCTGGCCTCCGCTCAGGAACGAGTTCAAGTACTTCCAAAGAATGACCACAACCTCTTCAGTGGAAGGTAAACAGAATCTGGTGATTATGGGTAGGAAAACCTGGTTCTCCATTCCTGAGAAGAATCGACCTTTAAAGGACAGAATTAATATAGTTCTCAGTAGAGAACTCAAAGAACCACCACGAGGAGCTCATTTTCTTGCCAAAAGTTTGGATGATGCCTTAAGACTTATTGAACAACCGGAATTGGCAAGTAAAGTAGACATGGTTTGGATAGTCGGAGGCAGTTCTGTTTACCAGGAAGCCATGAATCAACCAGGCCACCTTAGACTCTTTGTGACAAGGATCATGCAGGAATTTGAAAGTGACACGTTTTTCCCAGAAATTGATTTGGGGAAATATAAACTTCTCCCAGAATACCCAGGCGTCCTCTCTGAGGTCCAGGAGGAAAAAGGCATCAAGTATAAGTTTGAAGTCTACGAGAAGAAAGAC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein\n",
       "---------------------------------------------------------------------\n",
       "Stats:\n",
       "    length: 186\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    has stops: False\n",
       "---------------------------------------------------------------------\n",
       "0   VRPLNCIVAV SQNMGIGKNG DLPWPPLRNE FKYFQRMTTT SSVEGKQNLV IMGRKTWFSI\n",
       "60  PEKNRPLKDR INIVLSRELK EPPRGAHFLA KSLDDALRLI EQPELASKVD MVWIVGGSSV\n",
       "120 YQEAMNQPGH LRLFVTRIMQ EFESDTFFPE IDLGKYKLLP EYPGVLSEVQ EEKGIKYKFE\n",
       "180 VYEKKD"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_dna_translation = skbio.DNA(wt_dna).translate()\n",
    "wt_dna_translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure that the translated protein matches wildtype\n",
    "np.all(wt_dna_translation.values == wt_DHFR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "class EnergyFunctionCalculator:\n",
    "    \n",
    "    def __init__(self, \n",
    "                 h_i_a, # local fields\n",
    "                 e_i_j_a_b, # pairwise couplings\n",
    "                 alphabet_order = AMINO_ACIDS):\n",
    "        \"\"\"Set the local fields and pairwise couplings\"\"\"\n",
    "        self.h_i_a = h_i_a\n",
    "        self.e_i_j_a_b = e_i_j_a_b\n",
    "        self.alphabet_order = alphabet_order\n",
    "        self.N = h_i_a.shape[0] # length of the protein\n",
    "        # seq_runner is just a sequential array from 0 to the length of the protein\n",
    "        # if it used as an index in the calculations. \n",
    "        self.seq_runner = np.array(range(self.N), np.int)\n",
    "    \n",
    "    def __call__(self, protein):\n",
    "        \"\"\"Take an protein sequence and calculate the energy function from the evolutionary MSA\"\"\"\n",
    "        if protein.has_gaps():\n",
    "            raise ValueError(\"Protein Sequence has gaps in it\")\n",
    "        if protein.has_stops():\n",
    "            raise ValueError(\"Protein Sequence has stops in it\")\n",
    "        N = protein.values.size\n",
    "        if N != self.N:\n",
    "            raise ValueError(\"Protein Sequence length {} and Energy Calculator length {} do not match\".format(N, self.N))\n",
    "        idx = np.zeros(protein.values.size, dtype=np.uint16)\n",
    "        for i, a in enumerate(self.alphabet_order):\n",
    "            idx[protein.values == a] = i\n",
    "        idx = idx.squeeze()\n",
    "        # meshgrid creates a \"tuple\" product of the indices by repeating the index\n",
    "        # along the x axis in xv and repeating it along the yaxis in yv\n",
    "        xv, yv = np.meshgrid(idx, idx)\n",
    "        energy = -np.sum(self.e_i_j_a_b[self.seq_runner, xv, self.seq_runner, yv ]) / 2. - \\\n",
    "            np.sum(self.h_i_a[self.seq_runner, idx])\n",
    "        return(energy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "588711.3156041729"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein = wt_dna_translation\n",
    "energy_calc = EnergyFunctionCalculator(h_i_a, e_i_j_a_b, AMINO_ACIDS)\n",
    "energy_calc(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Mutants Energy Calculations <a id=\"mutants_energy\" />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Mutant:\n",
    "    \"\"\" This class stores the mutations from wildtype \"\"\"\n",
    "    wt = None # DNA or RNA or Protein sequence of wildtype \n",
    "    aa_chars = None # amino acid characters\n",
    "    \n",
    "    def __init__(self, idx = np.array([], dtype=np.uint16), \n",
    "                       values = np.array([], dtype='|S1')):\n",
    "        self.idx = idx\n",
    "        self.values = values\n",
    "        if self.idx.size != self.values.size:\n",
    "            raise ValueError('mutant index and values array have different sizes')\n",
    "    \n",
    "    def apply_mutations(self):\n",
    "        \"\"\" Apply the mutations for this mutant to wild-type and \n",
    "            return the new sequence of the same biological type as WT\"\"\"\n",
    "        mutant_seq = Mutant.wt.values.copy()\n",
    "        mutant_seq[self.idx] = self.values\n",
    "        return(type(Mutant.wt)(mutant_seq))\n",
    "\n",
    "Mutant.wt = skbio.DNA(wt_dna)\n",
    "Mutant.aa_chars = AMINO_ACIDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = Mutant(np.array([10,20,30], dtype=np.uint16), np.array(['G', 'A', 'T'], dtype='S1'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein\n",
       "---------------------------------------------------------------------\n",
       "Stats:\n",
       "    length: 186\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    has stops: False\n",
       "---------------------------------------------------------------------\n",
       "0   VRPWNCIVAV SQNMGIGKNG DLPWPPLRNE FKYFQRMTTT SSVEGKQNLV IMGRKTWFSI\n",
       "60  PEKNRPLKDR INIVLSRELK EPPRGAHFLA KSLDDALRLI EQPELASKVD MVWIVGGSSV\n",
       "120 YQEAMNQPGH LRLFVTRIMQ EFESDTFFPE IDLGKYKLLP EYPGVLSEVQ EEKGIKYKFE\n",
       "180 VYEKKD"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.apply_mutations().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "588994.3534789148"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "energy_calc(x.apply_mutations().translate())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulate rounds of neutral evolution <a id=\"simulate_evo\" />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "np.random.seed(1)\n",
    "\n",
    "DNA_ALPHABET = np.array(['A', 'C', 'T', 'G'], dtype=\"S1\")\n",
    "# Let's set the transition probabilities so that the \n",
    "# transition/transversion ratio is 2. \n",
    "transition_probabilities = {b'A' : np.array([0, 0.25, 0.25, 0.50]),\n",
    "                            b'C': np.array([0.25, 0, 0.50, 0.25]),\n",
    "                            b'T': np.array([0.25, 0.50, 0, 0.25]),\n",
    "                            b'G': np.array([0.50, 0.25, 0.25, 0])}\n",
    "\n",
    "# size of the DNA sequence\n",
    "DNA_L = Mutant.wt.values.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "evolution_round_size = 10000\n",
    "mutant_list = [None] * evolution_round_size\n",
    "mutant_energies = np.zeros(evolution_round_size, np.float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import clear_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing mutant number 10000 \n"
     ]
    }
   ],
   "source": [
    "# create new mutants and calculate their energy function\n",
    "    \n",
    "count = 0    \n",
    "while count < evolution_round_size:\n",
    "    # make an average of 4 DNA changes per sequence according to the binomial distribution\n",
    "    num_changes = np.random.binomial(DNA_L, 4./DNA_L)\n",
    "    # Uniformly choose which indices will be mutatbed\n",
    "    idxs_to_change = np.random.choice(DNA_L, num_changes, replace=False)\n",
    "    values_to_change = Mutant.wt.values[idxs_to_change]\n",
    "\n",
    "    changed_values = values_to_change.copy()\n",
    "\n",
    "    # Find which nucleotides need changing and change them. \n",
    "    for nucleotide in DNA_ALPHABET:\n",
    "        # find the indices that match the current nucleotide\n",
    "        nuc_indices = np.where(values_to_change == nucleotide)[0]\n",
    "        new_nucleotides = np.random.choice(DNA_ALPHABET, size=nuc_indices.size, \n",
    "                                           replace=True, \n",
    "                                           p=transition_probabilities[nucleotide])\n",
    "        changed_values[nuc_indices] = new_nucleotides\n",
    "        mutant = Mutant(idxs_to_change, changed_values)\n",
    "        try:  \n",
    "            energy = energy_calc(mutant.apply_mutations().translate())\n",
    "        except ValueError: # if there are stops, gaps etc etc\n",
    "            pass\n",
    "        else: # we will only add to our list if the energy_calc doesn't throw an exception\n",
    "            if (count < evolution_round_size):\n",
    "                mutant_list[count] = mutant\n",
    "                mutant_energies[count] = energy\n",
    "            count += 1\n",
    "            if count % 100 == 0:\n",
    "                clear_output(wait=True)\n",
    "                print(\"Processing mutant number {} \".format(count))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEWCAYAAAB8LwAVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3Xm4HGWZ9/Hvj4QlLIYlkSVBAh4WM76okEEUfHEXxAMOIJJXdgiDDo446hgRlblcBpVxFAUhKIugrCISA8MixoiyBQUDQobIMiEBCUsSEtBAuN8/6jmkOFPdp85SXd3n/D7X1depfqq66n6q+tRd9TzVVYoIzMzMelur7gDMzKw9OUGYmVkhJwgzMyvkBGFmZoWcIMzMrJAThJmZFXKC6BCS7pX09rrjqJOkf5C0UNIKSW+qO55mJIWkrrrjqJOkj0r6S9pem7VwuSdJ+kGrljecOUG0AUkPS3p3r7IjJd3c8z4i/i4iZvcxn0lpxzS6olDrdhpwQkRsGBF/6D0y1X1l2iEtkvQtSaNqiLMpSQdL+p2k5yTNrjueKkhaG/gW8N60vZ6qaDlvl/RoviwivhYRx1axvJHGCcJKa4PEsw1wbx/TvCEiNgT2Aj4MHF15VP33NPBt4NS6A+lRwbbdHFiPvreXtTEniA6RP8uQtJukuZKWp1P4b6XJ5qS/S9NR9FskrSXpZEmPSHpC0o8kjc3N9/A07ilJX+i1nFMkXSHpIknLgSPTsm+RtFTSY5K+J2md3PxC0sckPSDpWUlflvTadMS8XNJl+el71bEwVknrSloBjALulvTnvtZXRCwAfgu8MTf/rSRdLelpSQskTcuNO1/SV3LvX3FkmtbLpyX9UdIySZdKWi83/jNpfSyW1DQpRcSNEXEZsLiveqR5f0DSXWmd/07Szv2Iq6/PflbSH4GVkkZL2kXSH9K2uzzN7ytp+nskdec+v7akJ9WruU/SDsD89HappJuKzm4lzZZ0bBo+UtLNkk6T9IykhyTtk5t2U0nnpfX7jKSrJG0AXAtslb7vK9I2PkXSRbnP7qesiXZpWubryqw/SeMk/SJ97mlJv5E0svaZEeFXzS/gYeDdvcqOBG4umga4BTgsDW8I7J6GJwEBjM597mhgAbBdmvZK4MI0bjKwAtgTWIesCeeF3HJOSe8/SHYwMQbYFdgdGJ2Wdx9wYm55AfwceBXwd8DfgF+m5Y8F/gQc0WA9NIw1N++uJuvx5fHATsBjwCdz4+cAZ5Id2b4RWAK8M407H/hKbtq3A4/2Wv+3A1sBm6Z6H5/G7Q38BXg9sAHwk75iTZ87FpjdxzRvAp4A3kyWII9IsaxbIq4yn70L2Dpt23WAR4BPAGsDBwCretYL8K/ApbnY9gfmNYh7ErnvYu/3qWw2cGzu+/4CMC3F+lGyBKo0fhZwKbBJim2vou2U+95elIZ3AFYC70mf+1ey79g6JdbfvwNnpc+tDbytJ56R8hpZ2bC9XZWOVJZKWkq2I2vkBaBL0riIWBERtzaZ9iPAtyLiwYhYAXwOOCQdyR0EzIyImyNiFfBFsn/ivFsi4qqIeCkino+IOyPi1oh4MSIeBs4ma87J+0ZELI+Ie4F7gOvT8peRHfE16mBuFmtZv5e0kuwffTZpPUraGtgD+GxE/DUi7gJ+ABzej3mfHhGLI+JpYCZrzk4OBs6LiHsiYiXZDmqoHAecHRG3RcTqiLiALOnuXiKusp9dGBHPsybxnx4RL0TElWQ7zx4XAe+X9Kr0/jDgwiGs6yMRcU5ErAYuALYENpe0JbAP2Y77mRTbr0vO88PArIi4ISJeIDsIGgO8NTdNo/X3Qophm7TM30TKHCOFE0T7+GBEbNzzAj7WZNpjyI6M7pd0h6QPNJl2K7Kjwh6PkO0ENk/jFvaMiIjngN6diQvzbyTtkE67H1fW7PQ1YFyvz/wlN/x8wfsNBxBrWbuk+X+Y7Mh5g9y8n46IZ3vNf0I/5v14bvg51tTjFeuRV9ZhsLYBPtXr4GHrtMy+4irz2XzcWwGLeu0E89+PxWTNdgdK2phsp/3jwVXvFV6uR/ouQlaXrcm23TMDmOcrvlMR8RJZnfLbvdH6+ybZ2cb1kh6UNH0Ay+9oThAdKCIeiIipwKuBrwNXpPbYoqObxWQ7ih6vAV4k22k/BkzsGSFpDND7csTe8/w+cD+wfUS8CjgJ0MBrUzrW0iJzGVlT3Bdz895U0ka95r8oDa8E1s+N26Ifi3yMbCeWn+9QWQh8NX/wEBHrR8TFQ/TZ/PZ9DJggKb898/WC7Mj+UOBDZGeXiyhnZfo7kHW8kGzbbVwwrq8j+ld8p1LdtmbNdm8oIp6NiE9FxHbAfsC/SHpXyZiHBSeIDiTpUEnj09HQ0lT8Elmb+ktkbfg9LgY+KWlbSRuSHfFfGhEvAlcA3ZLeqqzj+BT63tlvBCwHVkjaiayteKg0i3UgTgWmSdoiIhYCvwP+XdJ6qbP2GLJmE8ja4t+fOkO3AE7sx3IuI+vAnyxpfeBLzSaWNCp1hI4G1krxrN1g8nOA4yW9WZkNJO3bK9E10t/P3gKsBk5IHdb7A7v1muYqsrO0TwA/KhEDABGxhGynfGiq/9HAa0t+9jGypskzJW2SOsf/bxr9F2Az5S686OUyYF9J70rr+FNkzWy/62u5yjr4u1JSWUa2bl4qE/Nw4QTRmfYG7lV2Zc93gENS/8BzwFeB36Ymhd2Bc8naiecADwF/BT4OkPoIPg5cQnb0uIKsU/NvTZb9aeD/Ac+S7YAuHcJ6NYx1ICJiXprXZ1LRVLLO0sXAz4AvRcSNadyFwN1knZbX0496RcS1ZJet3kTWJHFTHx85jKyp7ftkHZ/Pk63LonnPJeu4/R7wTJr/kSXj6tdnUz/UAWSJcynZmcIvyH0fUl/FT4FtyS4i6I9pZNviKbILGPrcSeccRtYncD/Zd/TEFM/9ZAcWD6bvfL75jIiYn+rxXeBJoBvoTnXty/bAjWT/F7cAZ0bEr/oRc8fruULAjHTUvpSs+eihuuOx+km6DTgrIs7LlX0R2CEiDq0vMmsFn0GMcJK6Ja2f+jBOA+aRHUXbCCRpL0lbpCamI4Cdgf/Kjd+U7AxjRl0xWus4Qdj+ZE0ui8lOqQ8ZaZfy2SvsSNbUtpSsvf6g1AeAsh8WLgSujYg5jWdhw4WbmMzMrJDPIMzMrFDdN18blHHjxsWkSZPqDsPMrKPceeedT0bE+L6m68gEoeyGYd1dXV3MnTu37nDMzDqKpFK/9u/IJqaImBkRx40d2+i3MWZmNlgdmSDMzKx6ThBmZlbICcLMzAo5QZiZWSEnCDMzK+QEYWZmhZwgzMyskBOEmZkV6sgEkW5RPWPZsmV1h2JmNmx1ZILwL6nNzKrXkQnCzMyq5wRhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JtkyAkvU7SWZKukPTRuuMxMxvpKk0Qks6V9ISke3qV7y1pvqQFkqYDRMR9EXE8cDCwR5VxmZlZ36o+gzgf2DtfIGkUcAawDzAZmCppchq3HzALuKbiuMzMrA+VJoiImAM83at4N2BBRDwYEauAS4D90/RXR8Q+wEcazVPScZLmSpq7ZMmSqkI3MxvxRtewzAnAwtz7R4E3S3o7cACwLk3OICJiBjADYMqUKVFdmGZmI1sdCaJQRMwGZtcchpmZJXVcxbQI2Dr3fmIqK01St6QZy5YtG9LAzMxsjToSxB3A9pK2lbQOcAhwdX9mEBEzI+K4sWPHVhKgmZlVf5nrxcAtwI6SHpV0TES8CJwAXAfcB1wWEfdWGYeZmfVfpX0QETG1Qfk1DOJSVkndQHdXV9dAZ2FmZn1om19S94ebmMzMqteRCcLMzKrXkQnCVzGZmVWvIxOEm5jMzKrXkQnCzMyq5wRhZmaFOjJBuA/CzKx6HZkg3AdhZla9jkwQZmZWPScIMzMr5ARhZmaFOjJBuJPazKx6HZkg3EltZla9jkwQZmZWPScIMzMr5ARhZmaFnCDMzKxQRyYIX8VkZla9jkwQvorJzKx6HZkgzMysek4QZmZWyAnCzMwKjW42UtJbgEOBtwFbAs8D9wCzgIsiwr3EZmbDVMMzCEnXAscC1wF7kyWIycDJwHrAzyXt14ogzcys9ZqdQRwWEU/2KlsB/D69/kPSuMoia0JSN9Dd1dVVx+LNzEaEhmcQ+eQgaRtJ707DYyRt1HuaVvJlrmZm1euzk1rSNOAK4OxUNBG4qsqgzMysfmWuYvonYA9gOUBEPAC8usqgzMysfmUSxN8iYlXPG0mjgaguJDMzawdlEsSvJZ0EjJH0HuByYGa1YZmZWd3KJIjpwBJgHvCPwDVkl7qamdkw1vSHcgAR8RJwTnqZmdkI0TBBSJpHk76GiNi5kojMzKwtNDuD+EDLojAzs7bTMEFExCOtDMTMzNpLmR/K7S7pDkkrJK2StFrS8lYE1yQmP1HOzKxiZa5i+h4wFXgAGEN2A78zqgyqL77VhplZ9Uo9DyIiFgCjImJ1RJxHdndXMzMbxvq8zBV4TtI6wF2SvgE8hh80ZGY27JXZ0R+WpjsBWAlsDRxYZVBmZla/MmcQTwKrIuKvwL9JGgWsW21YZmZWtzJnEL8E1s+9HwPcWE04ZmbWLsokiPUiYkXPmzS8fpPpzcxsGCiTIFZK2qXnjaRdgeerC8nMzNpBmT6IE4HLJS0GBGwBfLjSqMzMrHZl7uZ6h6SdgB1T0fyIeKHasMzMrG5lbrXxIbJ+iHuADwKX5puczMxseCrTB/GFiHhW0p7Au4AfAt+vNiwzM6tbmQSxOv3dFzgnImYB61QXkpmZtYMyCWKRpLPJOqavkbRuyc+ZmVkHK7OjPxi4DnhfRCwFNgU+U0Uwkj4o6RxJl0p6bxXLMDOzcvpMEBHxXERcGREPpPePRcT1ZRcg6VxJT0i6p1f53pLmS1ogaXqa91URMQ04Hl9Ka2ZWq1Y0FZ1Pr9uDp/s5nQHsA0wGpkqanJvkZGp+5oT136Tps2r5rJlVo/IEERFzgKd7Fe8GLIiIByNiFXAJsL8yXweujYjfF81P0nGS5kqau2TJkmqDNzMbwerqbJ4ALMy9fzSVfRx4N3CQpOOLPhgRMyJiSkRMGT9+fPWRmpmNUGV+KPespOW9Xgsl/UzSdkMZTEScHhG7RsTxEXFWk5j8TOoWcLOP2chW5gzi22RXLU0AJgKfBn5C1ix07gCXu4jswUM9JqayUvxM6noMNmE44Zh1ljIJYr+IODsino2I5RExg+yS10uBTQa43DuA7SVtmx5neghw9QDnZcOYk4pZfcokiOckHSxprfQ6GPhrGhd9fVjSxcAtwI6SHpV0TES8SPYI0+uA+4DLIuLeAdbBKtRsB13lGUVf83biMKtemQTxEbLnUj+RXocBh0oaQ7aTbyoipkbElhGxdkRMjIgfpvJrImKHiHhtRHy1P0G7D6J+dezAnRTMWqvMD+UejIjuiBiXXt0RsSAino+Im1sRZEFM7oNoM4123kXl3tGbdYYyVzFNTFcsPZFeP5U0sRXB2cjmRGJWrzJNTOeRdSBvlV4zU1lt3MTUubzTN+scZRLE+Ig4LyJeTK/zgVp/oeYmpvoMZAfvpGDWmcokiKckHSppVHodCjxVdWDWmfLJoL+JoT/9GGZWvTIJ4miyW34/DjwGHAQcVWVQ1t4GkwTMrHOM7muCiHgE2K8FsZQmqRvo7urqqjuUYWnS9Fk8fOq+dYdRyAnJrHUaJghJ36XJD+Ei4p8riaiEiJgJzJwyZcq0umKw6jgJmLWHZmcQc1sWhbWNVu2cW5kE2vmMyKydNUwQEXFBKwOxenXqTnSgcXdqfc1aqWEndXo29OsbjNtA0tGSPlJdaFa1skfxVR7t9563r2Qyax/NrmI6A/iipPskXS7pzPR86d8AvwM2Aq5oSZS9+Idy1RvqHXIr7s3kJGI2tJo1Md0FHCxpQ2AKsCXwPHBfRMxvUXyNYnMndQUGuoP1jtpseCpzs74VETE7Ii6OiKvqTg5Wj07Z6XdKnGadoK5nUpu1nJOHWf84QdiwNGn6LCcEs0Eqc7vv/9OKQPrDndRDyztSMytS5gziTEm3S/qYpLa4farv5mrNOOGZDY0yndRvI3vs6NbAnZJ+Iuk9lUdmI0ard+hOIGbllOqDiIgHgJOBzwJ7AadLul/SAVUGZ2Zm9SnTB7GzpP8E7gPeCXRHxOvS8H9WHJ9VzEfTZtZIn7f7Br4L/AA4KSKe7ymMiMWSTq4sMrMh4ARoNnBlEsS+wPMRsRpA0lrAehHxXERcWGl0ZmZWmzJ9EDcCY3Lv109lZmY2jJVJEOtFxIqeN2l4/epC6pt/BzF4bnop5vVitkaZBLFS0i49byTtSnbTvtr4dxADM1x3fkP1q+nhun7MBqpMgjgRuFzSbyTdDFwKnFBtWGZDq9nO34nBrFifndQRcYeknYAdU9H8iHih2rDMzKxuZW/W9/fAzsAuwFRJh1cXkg1W/ojYz2p4pcHUf6SvOxt5yvxQ7kLgNGBPskTx92QPEDLraN7hmzVX5ncQU4DJERFVB2NmZu2jTBPTPcAWVQdiZmbtpUyCGAf8SdJ1kq7ueVUdmFXHTStrNFoX/V1HXqc2HJVpYjql6iCset6BmVl/lbnM9deStgG2j4gbJa0PjKo+tMYkdQPdXV1ddYZhZjaslbmKaRpwBXB2KpoAXFVlUH3xL6nNzKpXpg/in4A9gOXw8sODXl1lUGZmVr8yCeJvEbGq542k0YAvee0A7ncob6g6q82GkzIJ4teSTgLGpGdRXw7MrDYsMzOrW5kEMR1YAswD/hG4huz51GbWi884bDjpM0FExEsRcU5EfCgiDkrDbmKyEWOwO30nDetUfV7mKukhCvocImK7SiIyM7O2UPZeTD3WAz4EbFpNOGZm1i7KNDE9lXstiohvA/u2IDYzM6tRmR/K7ZJ7TZF0POXOPKzFmj0HwlprqNa/t6PVqcyO/j9ywy8CDwMHVxKNmZm1jTJNTO/Ivd4TEdMiYn4rgjOrw0CP2os+V7ZssMs2q0KZq5j+pdn4iPjW0IVjZmbtoswP5aYAHyW7Sd8E4HiyZ1NvlF7W5nxUOng969Dr0kaSMn0QE4FdIuJZAEmnALMi4tChDETSdsDngbERcdBQztvMzPqvzBnE5sCq3PtVqaxPks6V9ISke3qV7y1pvqQFkqYDRMSDEXFM2cDNOk1/rjLzmYq1gzIJ4kfA7ZJOSWcPtwEXlJz/+cDe+QJJo4AzgH2AycBUSZPLBmxmZq1R5iqmrwJHAc+k11ER8bUyM4+IOcDTvYp3AxakM4ZVwCXA/mUDlnScpLmS5i5ZsqTsx4atSdNn+fcPLVZ2fZfdFr23oVm7KHMGAbA+sDwivgM8KmnbQSxzArAw9/5RYIKkzSSdBbxJ0ucafTgiZkTElIiYMn78+EGEYWZmzZS5zPVLZFcy7QicB6wNXET2lLkhExFPkV0hZWZmbaDMGcQ/APsBKwEiYjGDu7x1EbB17v3EVFaapG5JM5YtWzaIMDqfmyWq5fVrI12ZBLEqPf8hACRtMMhl3gFsL2lbSesAhwBX92cGETEzIo4bO3bsIEMxM7NGyiSIyySdDWwsaRpwI3BOmZlLuhi4BdhR0qOSjomIF4ETgOuA+4DLIuLegYVvZmZV6bMPIiJOS8+iXk7WD/HFiLihzMwjYmqD8mvIHl06IJK6ge6urq6BzqLjTJo+i4dPbXyXdTeHmNlQa5og0m8WboyIdwClkkIrRMRMYOaUKVOm1R2Lmdlw1bSJKSJWAy9JcmO/mdkIU+ZeTCuAeZJuIF3JBBAR/1xZVGZmVrsyndRXAl8A5gB35l61GSmXufbuV3A/Q+cb7Db0d8BaqeEZhKTXRMT/RETZ+y61jPsgzMyq1+wM4qqeAUk/bUEsZmbWRpolCOWGt6s6EDMzay/NEkQ0GK7dcOqDcJty5xuqbeg+J2s3zRLEGyQtl/QssHMaXi7pWUnLWxVgEd9qw8yseg07qSNiVCsDMTOz9lL2eRBmZjbCdGSCGE59EL2VaXd223R7aLQdqtg+A1mWvyc2WB2ZINwHYWZWvY5MEGZmVj0nCDMzK+QEYWZmhZwgzMysUEcmiOF2FVNfv6DNv/eVKe2hv9thMNut6LOTps96uXwovxP+flleRyYIX8VkZla9jkwQZmZWPScIMzMr5ARhZmaFnCDMzKyQE4SZmRXqyATRKZe5VnkppC9H7HwDvdnfUN6gr+z0ZZbZykt/rTU6MkH4Mlczs+p1ZIIwM7PqOUGYmVkhJwgzMyvkBGFmZoWcIMzMrJAThJmZFXKCMDOzQk4QZmZWyAnCzMwKdWSCGIpbbdTxM/+BPBnOtyMY3oZq+w7kCXNlpmv0NLs61b38dtCqddCRCcK32jAzq15HJggzM6ueE4SZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCo2uO4AekjYAzgRWAbMj4sc1h2RmNqJVegYh6VxJT0i6p1f53pLmS1ogaXoqPgC4IiKmAftVGZeZmfWt6iam84G98wWSRgFnAPsAk4GpkiYDE4GFabLVFcdlZmZ9qDRBRMQc4OlexbsBCyLiwYhYBVwC7A88SpYkmsYl6ThJcyXNXbJkSRVhv0Lvp3Q1esJW/olefc2nzFO6/NQs620g34mi722z73B/pmk2XSsN9Ml4VS1rOKmjk3oCa84UIEsME4ArgQMlfR+Y2ejDETEjIqZExJTx48dXG6mZ2QjWNp3UEbESOKruOMzMLFPHGcQiYOvc+4mprDRJ3ZJmLFu2bEgDMzOzNepIEHcA20vaVtI6wCHA1f2ZQUTMjIjjxo4dW0mAZmZW/WWuFwO3ADtKelTSMRHxInACcB1wH3BZRNxbZRxmZtZ/lfZBRMTUBuXXANcMdL6SuoHurq6ugc7CzMz60JG32nATk5lZ9ToyQZiZWfU6MkH4KiYzs+opIuqOYcAkLQEeqXgx44AnK15GXVy3zuS6daZ2qts2EdHnL407OkG0gqS5ETGl7jiq4Lp1JtetM3Vi3TqyicnMzKrnBGFmZoWcIPo2o+4AKuS6dSbXrTN1XN3cB2FmZoV8BmFmZoWcIMzMrNCwThCSHpY0T9JdkuamsjdKurWnTNJuqXyspJmS7pZ0r6SjcvM5QtID6XVErnzXNP8Fkk6XpFS+qaQb0vQ3SNqkjeu2Ok1/l6Src+XbSrot1e3SdOddJK2b3i9I4yfVXLdNJP1M0h8l3S7p9bn5FD37fLjU7X/NJ5UXfveUOT3V7Y+SdmlR3d4g6ZZUPlPSq3LTfy7FM1/S+3LlnbLdBlK3tttuTUXEsH0BDwPjepVdD+yTht8PzE7DJwFfT8PjyR6Vug6wKfBg+rtJGt4kTXc7sDsg4NrcfL8BTE/D03vm2251S+9XNJj/ZcAhafgs4KNp+GPAWWn4EODSmuv2TeBLaXgn4JdpeBTwZ2C7tB3vBiYPh7o1mk+z716a77Xpu7o7cFuL6nYHsFcaPhr4chqenLbJusC2aVuN6rDt1q+6tet2a/Ya1mcQDQTQk+nHAotz5RtJErAh2U70ReB9wA0R8XREPAPcAOwtaUvgVRFxa2Rb8kfAB9O89gcuSMMX5Mqr1t+6FUrTvRO4IhXl65Cv2xXAu9L0VWtUt8nATQARcT8wSdLmNHj2+TCpWzONvnv7Az+KzK3Axuk7XLUdgDlp+AbgwFw8l0TE3yLiIWAB2TbrpO3W37o1027bDRjmTUxk/3jXS7pT0nGp7ETgm5IWAqcBn0vl3wNeR/bPOQ/4RES8RONnaE9Iw73LATaPiMfS8ONAX//UAzEUdQNYLzVr3Cqp50u5GbA0smd39K7by+sjjV+Wpq+rbncDBwCkppltyJ5S2Gi7DYe6NZoPNP7uNVofQ6kopnvJdnIAH2LN0ySb/V91ynbrb90azQfq3W4Ntc0zqSuyZ0QskvRq4AZJ9wMHAZ+MiJ9KOhj4IfBusjOFu8iOUl6bpv/NYAOIiJBUxbXEg65bRCwnuyfLIknbATdJmkf2D1an/tTtVOA7ku4iS35/AFbXFXgJQ1W3/zWfiJiTX1CF371Giup2NHC6pC+QPTlyVQvjGUpDVbd23G4NDesziIhYlP4+AfyM7DTvCODKNMnlrDn1Owq4Mp3KLQAeImv3bfQM7UWsOZrLlwP8pec0MP19YmhrNmR1y8/nQWA28CbgKbJT2Z4DiHzdXl4fafzYNH0tdYuI5RFxVES8ETicrI/lQRpvt+FQt0bzgcbfvUE/C34gdYuI+yPivRGxK3AxWXt8s3g6ZrsNoG5tud2aGbYJQtIGkjbqGQbeC9xD1syyV5rsncADafh/gHel6TcHdiT7Z7wOeK+yK0o2SfO5Lp0OLpe0e2rvPBz4eZrX1WT/9KS/PeVtVbdUp3VT+ThgD+BPqU/lV2RHtr3rkK/bQcBNafpa6iZpY6WrWYBjgTnpzKjw2efDoW5N5tO7Dr3rdnh2UYx2B5blmjQqq1s6UkbSWsDJZJ3LPfEcouwKpG2B7cku+uiY7dbfurXjdutTf3u1O+VFdhXE3el1L/D5VL4ncGcqvw3YNZVvRXY1yTyyjXZobl5Hk3U0LQCOypVPSdP+maydv+eX6ZsBvyT7R78R2LQd6wa8NZXdnf4e02sZt6c6Xw6sm8rXS+8XpPHb1Vy3twD/DcwnOwrfJDev96dxf+6Zz3CoW6P5NPvukV0Fc0ZaF/OAKS2q2ydSHf6brMlMuc98PsUzn3QVV4dtt37VrR23W18v32rDzMwKDdsmJjMzGxwnCDMzK+QEYWZmhZwgzMyskBOEmZkVGu6/pDYbEEmryS4r7HFJRJxaVzxmdfBlrmYFJK2IiA2HeJ6jY829hMzanpuYzPpB2f38/03S75Xd13+nVL6BpHOVPbfhD5L2T+VHSrpa0k3ALyWtJelMSfcru+//NZIOkvROSVfllvMeST+rqZpmgBOEWSNjtOZBSndJ+nBu3JMRsQvwfeDTqezzZLdxKrEtAAABW0lEQVR42A14B9ndWTdI43YBDoqIvcjuzjqJ7Fbeh5H9Whqy20jsJGl8en8UcG5FdTMrxX0QZsWej+wmeUV6bqx3J+l23GT31dlPUk/CWA94TRq+ISKeTsN7ApdHdrv1xyX9Cl6+g+eFwKGSziNLHIcPXXXM+s8Jwqz//pb+rmbN/5CAAyNifn5CSW8GVpac73nATOCvZEnE/RVWKzcxmQ2N64CPpzv7IulNDab7LXBg6ovYHHh7z4iIWEx2Z9eTyZKFWa18BmFWbIyyB/X0+K+ImN5k+i8D3wb+mG7//BDwgYLpfkp26/U/kT0p7Pe88gFNPwbGR8R9gwnebCj4MlezFpO0YUSskLQZ2e2p94iIx9O47wF/iIgf1hqkGT6DMKvDLyRtDKwDfDmXHO4k66/4VJ3BmfXwGYSZmRVyJ7WZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZof8PkdLRc8inZlYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "_ = plt.hist(mutant_energies, bins=500, log=True)\n",
    "plt.title('Histogram of Round 1 energy functions')\n",
    "plt.xlabel('Energy')\n",
    "_ = plt.ylabel('Frequency (log scale)')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
